src/Tutorials/Custom-Grain-Storage.md

```diff
@@ -0,0 +1,232 @@
+---
+layout: page
+title: Custom Grain Storage
+---
+
+# Custom Grain Storage
+
+## Writing a Custom Grain Storage
+
+In the tutorial on declarative actor storage, we looked at allowing grains to store their state in an Azure table using one of the built-in storage providers.
+While Azure is a great place to squirrel away your data, there are many alternatives.
+In fact, there are so many that there was no way to support them all.
+Instead, Orleans is designed to let you easily add support for your own form of storage by writing a grain storage.
+
+In this tutorial, we'll walk through how to write a simple file-based grain storage.
+A file system is not necessarily the best place to store data for grains, since it's so local, but it's an easy example to help us illustrate the principles.
+
+## Getting Started
+
+An Orleans grain storage is a class that implements `IGrainStorage` which is included in [Microsoft.Orleans.Core NuGet package](https://www.nuget.org/packages/Microsoft.Orleans.Core/).
+
+We also inherit from `ILifecycleParticipant<ISiloLifecycle>` which will allow us to subscribe to a particular event in the lifecycle of the silo.
+
+We start by creating a class named `FileGrainStorage`.
+
+```csharp
+using Orleans;
+using System;
+using Orleans.Storage;
+using Orleans.Runtime;
+using System.Threading.Tasks;
+
+namespace GrainStorage
+{
+    public class FileGrainStorage : IGrainStorage, ILifecycleParticipant<ISiloLifecycle>
+    {
+        private readonly string _storageName;
+        private readonly string _rootDirectory;
+        private readonly IGrainFactory _grainFactory;
+        private readonly ITypeResolver _typeResolver;
+        private JsonSerializerSettings _jsonSettings;
+
+        public FileGrainStorage(string storageName, string rootDirectory, IGrainFactory grainFactory, ITypeResolver typeResolver)
```

Comment:
Suggest passing the entire FileGrainStorageOptions to the implementation rather than just the root directory, that way options are automatically piped through if more options are added.

Reply:
Nice, I changed the constructor to take the options as argument.